tests/core/test_core_edm_copy.cpp

```diff
@@ -0,0 +1,279 @@
+
+    compare(vecmem::get_data(host), vecmem::get_data(buffer));
+}
+
+TEST_F(core_edm_copy_test, host_to_fixed_device_jagged) {
+
+    vecmem::edm::host<jagged_schema> host{m_resource};
+    fill(host);
+
+    vecmem::edm::buffer<jagged_schema> buffer{
+        std::vector<std::size_t>{2, 0, 3}, m_resource, nullptr,
+        vecmem::data::buffer_type::fixed_size};
+    m_copy.setup(buffer);
+
+    m_copy(vecmem::get_data(host), buffer, vecmem::copy::type::host_to_host);
+
+    compare(vecmem::get_data(host), vecmem::get_data(buffer));
+}
+
+TEST_F(core_edm_copy_test, host_to_resizable_device_simple) {
+
+    vecmem::edm::host<simple_schema> host{m_resource};
+    fill(host);
+
+    vecmem::edm::buffer<simple_schema> buffer{
+        10u, m_resource, vecmem::data::buffer_type::resizable};
+    m_copy.setup(buffer);
+
+    m_copy(vecmem::get_data(host), buffer, vecmem::copy::type::host_to_host);
+
+    compare(vecmem::get_data(host), vecmem::get_data(buffer));
+}
+
+TEST_F(core_edm_copy_test, host_to_resizable_device_jagged) {
+
+    vecmem::edm::host<jagged_schema> host{m_resource};
+    fill(host);
+
+    vecmem::edm::buffer<jagged_schema> buffer{
+        std::vector<std::size_t>{5, 5, 5}, m_resource, nullptr,
+        vecmem::data::buffer_type::resizable};
+    m_copy.setup(buffer);
+
+    m_copy(vecmem::get_data(host), buffer, vecmem::copy::type::host_to_host);
+
+    compare(vecmem::get_data(host), vecmem::get_data(buffer));
+}
+
+TEST_F(core_edm_copy_test, host_to_device_to_fixed_device_simple) {
+
+    vecmem::edm::host<simple_schema> host{m_resource};
+    fill(host);
+
+    vecmem::edm::buffer<simple_schema> buffer1{
+        static_cast<vecmem::edm::buffer<jagged_schema>::size_type>(host.size()),
+        m_resource, vecmem::data::buffer_type::fixed_size};
+    m_copy.setup(buffer1);
+
+    vecmem::edm::buffer<simple_schema> buffer2{
+        static_cast<vecmem::edm::buffer<jagged_schema>::size_type>(host.size()),
+        m_resource, vecmem::data::buffer_type::fixed_size};
+    m_copy.setup(buffer2);
+
+    m_copy(vecmem::get_data(host), buffer1, vecmem::copy::type::host_to_host);
+    m_copy(buffer1, buffer2, vecmem::copy::type::host_to_host);
+
+    compare(vecmem::get_data(host), vecmem::get_data(buffer2));
+}
+
+TEST_F(core_edm_copy_test, host_to_device_to_fixed_device_jagged) {
+
+    vecmem::edm::host<jagged_schema> host{m_resource};
+    fill(host);
+
+    vecmem::edm::buffer<jagged_schema> buffer1{
+        std::vector<std::size_t>{2, 0, 3}, m_resource, nullptr,
+        vecmem::data::buffer_type::fixed_size};
+    m_copy.setup(buffer1);
+
+    vecmem::edm::buffer<jagged_schema> buffer2{
+        std::vector<std::size_t>{2, 0, 3}, m_resource, nullptr,
+        vecmem::data::buffer_type::fixed_size};
+    m_copy.setup(buffer2);
+
+    m_copy(vecmem::get_data(host), buffer1, vecmem::copy::type::host_to_host);
+    m_copy(buffer1, buffer2, vecmem::copy::type::host_to_host);
```

Comment:
Should this test the error case where the sizes don't line up but `fixed_size` is given?

Reply:
I was thinking about it... :thinking: The reason I found the pre-existing bug in `vecmem::copy::set_sizes(...)` is that I made a typo with the sizes in `host_to_fixed_device_jagged`. And then the test failed "in the wrong way".

Okay, let me see how easily I could put in some simple tests where I expect the code to throw some kind of exception. :thinking:

Reply:
Adding the test also revealed that `vecmem::copy` was in some cases throwing `std::length_error` and in others `std::runtime_error` for these types of errors.

So I tried to fix those, and added some explicit tests for the SoA copies. That they would throw the expected exceptions. (Didn't add such tests for the lower level tests at this point though...)